qiskit_nature/mappers/second_quantization/bksf.py

```diff
@@ -0,0 +1,511 @@
+    b_q = edge_operator_bi(edge_list, q)
+    id_op = _pauli_id(edge_list.shape[1])
+    qubit_op = (id_op - b_p) * (id_op - b_q)
+    if p == s:  # two commutations
+        final_coeff = 0.25
+    else:  # one commutation
+        final_coeff = -0.25
+    qubit_op = (final_coeff * h2_pqrs) * qubit_op
+    return qubit_op
+
+
+## SW2018 eq 35
+## Includes contributions from a h.c. pair
+def _excitation_operator(
+    edge_list: List, p: int, q: int, h1_pq: float
+):  # pylint: disable=invalid-name
+    if p >= q:
+        raise ValueError("Expected p < q, got p = ", p, ", q = ", q)
+    b_a = edge_operator_bi(edge_list, p)
+    b_b = edge_operator_bi(edge_list, q)
+    a_ab = edge_operator_aij(edge_list, p, q)
+    qubit_op = (-1j * 0.5 * h1_pq) * ((b_b & a_ab) + (a_ab & b_a))
+    return qubit_op
+
+
+## SW2018 eq 37
+def _double_excitation(
+    edge_list: List, p: int, q: int, r: int, s: int, h2_pqrs: float
+):  # pylint: disable=invalid-name
+    b_p = edge_operator_bi(edge_list, p)
+    b_q = edge_operator_bi(edge_list, q)
+    b_r = edge_operator_bi(edge_list, r)
+    b_s = edge_operator_bi(edge_list, s)
+    a_pq = edge_operator_aij(edge_list, p, q)
+    a_rs = edge_operator_aij(edge_list, r, s)
+    a_pq = -a_pq if q < p else a_pq
+    a_rs = -a_rs if s < r else a_rs
+
+    id_op = _pauli_id(edge_list.shape[1])
+    qubit_op = (a_pq * a_rs) * (
+        -id_op
+        - b_p * b_q
+        + b_p * b_r
+        + b_p * b_s
+        + b_q * b_r
+        + b_q * b_s
+        - b_r * b_s
+        + b_p * b_q * b_r * b_s  ## Agrees with SW2018 eq 37 and OpenFermion. Aqua had `-`.
+    )
+    final_coeff = 0.125
+    qubit_op = (final_coeff * h2_pqrs) * qubit_op
+    return qubit_op
+
+
+def _number_excitation(
+    edge_list: List, p: int, q: int, r: int, s: int, h2_pqrs: float
+):  # pylint: disable=invalid-name
+    b_p = edge_operator_bi(edge_list, p)
+    b_q = edge_operator_bi(edge_list, q)
+    id_op = _pauli_id(edge_list.shape[1])
+    if p == r:
+        b_s = edge_operator_bi(edge_list, s)
+        a_qs = edge_operator_aij(edge_list, q, s)
+        a_qs = -a_qs if s < q else a_qs
+        qubit_op = (a_qs * b_s + b_q * a_qs) * (id_op - b_p)
+        final_coeff = 1j * 0.25
+    elif p == s:
+        b_r = edge_operator_bi(edge_list, r)
+        a_qr = edge_operator_aij(edge_list, q, r)
+        a_qr = -a_qr if r < q else a_qr
+        qubit_op = (a_qr * b_r + b_q * a_qr) * (id_op - b_p)
+        final_coeff = 1j * -0.25
+    elif q == r:
+        b_s = edge_operator_bi(edge_list, s)
+        a_ps = edge_operator_aij(edge_list, p, s)
+        a_ps = -a_ps if s < p else a_ps
+        qubit_op = (a_ps * b_s + b_p * a_ps) * (id_op - b_q)
+        final_coeff = 1j * -0.25
+    elif q == s:
+        b_r = edge_operator_bi(edge_list, r)
+        a_pr = edge_operator_aij(edge_list, p, r)
+        a_pr = -a_pr if r < p else a_pr
+        qubit_op = (a_pr * b_r + b_p * a_pr) * (id_op - b_q)
+        final_coeff = 1j * 0.25
```

Comment:
Yes. I was thinking the same thing. Many of your comments are on code that I copied in from the old version. But, I did not get around to refactoring/commenting/clarifying everything yet.... oh btw. Everything like `a_pr * b_r` is deprecated in favor of `b_r & a_pr`. This should probably be fixed before release.

Reply:
I tried to factor out some of the computation. But, this adds to complexity and obscures the relation with the mathematical expressions. I agree this looks repetitious, but factoring out the computation is marginally worse, I think. Maybe someone else wants to weigh in.

Reply:
I think here makes sense not to extract the functionality so that one can read and follow the reference of the mapping. I think this duplication of code increase readability.